### DISCUSSION TITLE
Constants in Spin Hall Conductivity

### DISCUSSION THREAD

**Discussion by ma455173220** [wannier-berri/wannier-berri#283] [Q&A] [answered]
Dear Dr. Stepan,

I apologize for reaching out again, but I have a couple of questions related to the spin Hall conductivity in the code that I would greatly appreciate your insights on.

1. Constant "2" in the Formula: In Qiao's paper, we have the full spin Hall conductivity expressed in a formula expressed as: 
![image](https://github.com/wannier-berri/wannier-berri/assets/42956329/4af6f5d1-2866-489c-843e-770acf2df740)
In the WannierBerri code, the constants ![image](https://github.com/wannier-berri/wannier-berri/assets/42956329/66e36221-46d5-4817-bc5b-9d00190304a8) appear to be taken into account by the variable "self.constant_factor" in https://github.com/wannier-berri/wannier-berri/blob/4be040969e8df4817cb25f64a32784ba4e862e3d/wannierberri/calculators/dynamic.py#L86. Additionally, the two minus signs are canceled by each other, and hbar^2 is canceled by two (1/hbar) constants in the spin current operator and velocity operator. However, there remains a constant "2" in the formula that I couldn't identify where it is considered in the code. Could you please clarify the handling of this constant in the code?
2. Division of factor_omega by 2: In the code (https://github.com/wannier-berri/wannier-berri/blob/4be040969e8df4817cb25f64a32784ba4e862e3d/wannierberri/calculators/dynamic.py#L239), I noticed a division of factor_omega by 2. I'm struggling to grasp why this division by two is implemented in this particular context. Could you please shed some light on the reason behind this division?
3. Spin Current Operator "J": In the calculation of the spin current operator "J," is the variable SS_H (https://github.com/wannier-berri/wannier-berri/blob/4be040969e8df4817cb25f64a32784ba4e862e3d/wannierberri/formula/covariant.py#L437) equivalent to "sigma_y" in the formula? Or is it "s_y" where s_y = (hbar/2)*sigma_y?
4. Multiplication of (hbar/-2e): In Qiao's paper, it mentions the multiplication of (hbar/-2e):
![image](https://github.com/wannier-berri/wannier-berri/assets/42956329/22c48955-f851-4d47-86a3-cb02efc321fa)
I'm curious whether in WannierBerri, we also consider this multiplication factor.

Your insights and clarifications on these questions would be immensely helpful for me to better understand the code and its calculations. Thank you for your time and assistance.

Regards,
Hongyang

**Comment by stepan-tsirkin** [comment]
Dear Hongyang @ma455173220 ,

Thank you for your interest and careful questioning the implemented equations. This helps us to check code for bugs. 
The equation actually implemented is rather the one from Ryoo's paper 

![image](https://github.com/wannier-berri/wannier-berri/assets/22172977/ce992333-6f05-4941-b892-5b13c4cc875b)
![image](https://github.com/wannier-berri/wannier-berri/assets/22172977/43ed5ce1-350b-4848-9073-e4b52e95f92a)

So, there is a difference of two f's here: 
https://github.com/wannier-berri/wannier-berri/blob/e7d9cb6a83322afba347fc9369eb94962ee35266/wannierberri/calculators/dynamic.py#L49-L50

The conversion between the equations is by interchanging the two band indices in summation and  using that $\mathrm{Im}[\langle n|j|m\rangle\langle m|v|n\rangle ]  = -\mathrm{Im}[\langle m|j|n\rangle\langle n|v|m\rangle ]$

2. The division by 2 in factor_omega comes from expanding the denominator as 

$\frac{1}{x^2-a^2} = \frac{1}{2x}\left(\frac{1}{-x-a}-\frac{1}{x-a}\right)$

As I sketched in the note:
[2023-09-27-Note-15-19-spin-hall.pdf](https://github.com/wannier-berri/wannier-berri/files/12763846/2023-09-27-Note-15-19-spin-hall.pdf)

3. The SS_H is the $sigma$, i.e. the interpolated expectation value of the Pauli matrix

4. In wanierberri eq 1 of ryoo (equivalent eq 1 of qiao) are implemen5ted directly, to get the conductivity in SI units (S/m) That paragraph of Qiao just describes  factor conversion between Spin Berry curvature and SHC, so it is not relevant to the implementation.


Does that altogether resolves confusion in the factor of 2 ?

**Comment by ma455173220** [comment]
Dear Dr. Stepan,

I apologize for my delayed response and thank you for providing the information. I'm sorry to say that my understanding of the mathematics is quite limited, so, in fact, things have become even more confusing for me. I'm now even struggling to grasp the concept of the variable "self.constant_factor" ($\frac{e^2}{\hbar} \frac{1}{V N_k^3}$) in:"

https://github.com/wannier-berri/wannier-berri/blob/4be040969e8df4817cb25f64a32784ba4e862e3d/wannierberri/calculators/dynamic.py#L86

Additionally, shouldn't the equation $\frac{1}{x^2-a^2}=\frac{1}{2 x}\left(\frac{1}{-x-a}-\frac{1}{x-a}\right)$ be $\frac{1}{x^2-a^2}=-\frac{1}{2 x}\left(\frac{1}{-x-a}-\frac{1}{x-a}\right)$?

Best Regards,
Hongyang

**Reply by stepan-tsirkin** [reply]
1. The concept of self.constant_factor is simple. A class DynamicCalculator has properties `Formula`, `constant_factor`, `factor_omega`, `factor_Efermi`  .  Then, it calculates the integral


$$ 
{\rm constant \textunderscore factor} \int \frac{d^3k}{(2\pi)^3} \sum_{ln}^{E_l\neq E_n} {\rm formula.trace  \textunderscore ln}(l,n) \times {\rm factor  \textunderscore  Efermi}(E_l, E_n, E_F)\times {\rm factor  \textunderscore omega}(E_l, E_n,\omega) 
$$

Discretization 
$$\int \frac{d^3k}{(2\pi)^3} \approx \frac{1}{N_k V_{cell}} \sum_{k}$$
is used 

2. Yes, You are right. I forgot the minus sign.   I wrote those comments quickly, consider them as a guide to obtain the result, not like smth correct to the last symbol.

**Comment by ma455173220** [comment]
Dear Dr. Stepan,

Thank you for your prompt response and for providing clarity on my previous questions. I appreciate your assistance.

I have another query related to the factor_omega formula. Given that I need to make slight manual modifications to the Kubo formula, it appears that utilizing the old factor_omega formula would better serve my needs. This is the old factor_omega formula in WannierBerri:
![image](https://github.com/wannier-berri/wannier-berri/assets/42956329/172249a4-16c3-4efe-9459-de7fb0bad0b4)

I was wondering does the current factor_omega formula represent $\frac{1}{\left( E1-E2 \right) ^2-(\hbar \omega +i\delta )^2}$ (where the "E1 - E2" coming from the berry connection in numerator is omitted here)? If so, I have some questions about the variables involved. Why do we have a variable "cfac2" in the formula, and why is there a "1j * np.pi * cfac2" in the final "cfac"?

Furthermore, if I intend to write a formula like $\frac{1}{\left( E1-E2 \right) \left( E2-E3 \right) -(\hbar \omega +i\delta )^2}$, how should I define delta_minus and delta_plus? Should they be defined as self.smear(E3 - E2 - self.omega) and self.smear(E1 - E2 - self.omega), respectively?

Your guidance on these matters would be greatly appreciated.

Best Regards,
Hongyang

**Reply by stepan-tsirkin** [reply]
The old factor_omega on your screenshot represents exactly what is written under the real part of cfac1

In fact, is self.smear is Lorentzian, it gives the same imaginary part .  Just as an option one can switch to a Gaussian smeqaring

In the same way as in the new factor_omega 

https://github.com/wannier-berri/wannier-berri/blob/23e1f70266480660cef9580f997bd6ff7ac67476/wannierberri/calculators/dynamic.py#L234-L239

The identity 

$$
\lim_{\delta\to 0} {\rm Im}\frac{1}{x+i\eta}  = -i\delta(x)
$$

is used here.

**Reply by stepan-tsirkin** [reply]
> Furthermore, if I intend to write a formula like 1(E1−E2)(E2−E3)−(ℏω+iδ)2, how should I define delta_minus and delta_plus? Should they be defined as self.smear(E3 - E2 - self.omega) and self.smear(E1 - E2 - self.omega), respectively? 

I do not think there is sauch option in this case, this denominator is not factorized as (smth1-omega)*(smth2*omega). So just evaluate the fraction as it is, keeping real and imaginary parts together

A more diffcult question is that Dynamilca calculator is designed to evaluate smth in the form given [above](https://github.com/wannier-berri/wannier-berri/discussions/283#discussioncomment-7163691), i.e. when transitions between 2 bands are considered.  If you have 3 bands involved, you need to redefine the `__init__` and `__call__` methods to handle 3 bands. 

Also the "Formula" should be something giving a 3-indices matrix element, i.e. to have a `trace_lmn` method or similar. If not a big secret, what equation are you implementing?

**Reply by ma455173220** [reply]
Dear Dr. Stephan,

I appreciate your prompt response. The equation I wish to incorporate is depicted as follows:
<img width="591" alt="image" src="https://github.com/wannier-berri/wannier-berri/assets/42956329/2f7f38aa-cda6-4441-8fde-03a9caf2bfc9">

My intended approach is to include three nested "for loops" within the __call__(self, data_K) method of the DynamicCalculator class. This appears to be the most straightforward solution that comes to mind. I do, however, have some reservations about the factor_omega component.

Best Regards,
Hongyang

**Reply by stepan-tsirkin** [reply, marked as answer]
In this expression En=Em, and there is no omega, nor imaginary part in denominator. So, it is basically a static calculator

The fraction (after sum over l) is a berry curvature matrix for degenerate states mn, right? So, it is given by the [Omega.nn](https://github.com/wannier-berri/wannier-berri/blob/23e1f70266480660cef9580f997bd6ff7ac67476/wannierberri/formula/covariant.py#L126-L168)

The expression is very similar to what implemented in the [AHC_Zeeman_spin](https://docs.wannier-berri.org/en/master/docs/calculators.html#wannierberri.calculators.static.AHC_Zeeman_spin) calculator 

https://github.com/wannier-berri/wannier-berri/blob/23e1f70266480660cef9580f997bd6ff7ac67476/wannierberri/calculators/static.py#L540-L550

And the product of spin and berry curvature is 

https://github.com/wannier-berri/wannier-berri/blob/23e1f70266480660cef9580f997bd6ff7ac67476/wannierberri/formula/covariant.py#L578-L581

The only difference is that it has a first derivative of fermi-dirac distribution function (fder=1), while in your expression it will be `fder=0`

**Reply by ma455173220** [reply]
Dear Dr. Stephan, 

I greatly appreciate your assistance! Indeed, that corresponds precisely to the berry curvature. I will carefully examine your suggestions. My ultimate inquiry pertains to this particular formula:
<img width="606" alt="image" src="https://github.com/wannier-berri/wannier-berri/assets/42956329/0ee31bc8-478a-4bc2-8c8c-08aa14e0a646">
Do you believe that I can employ three nested "for loops" along with the self.factor_omega and self.factor_Efermi functions directly in this scenario?

Best Regards,
Hongyang

**Reply by stepan-tsirkin** [reply]
I think  you can, if you want.

**Reply by ma455173220** [reply]
Dear Dr. Stepan,

I truly appreciate your help and support!

Best Regards,
Hongyang
